DeltaFIFO cleanup

Adirio opened the pull request:
**What type of PR is this?**
/kind cleanup
/sig api-machinery

**What this PR does / why we need it**:
It does a clean up of some parts of the `DeltaFIFO` code:

1. A conditional that could never happen was being considered. This case checks for 0-length slices of Deltas. This slice is formed by adding a Delta to a previous slice, so it will always have at least one. After that the slice is deduped, but this method will never reduce the length to zero unless it was already zero. The behavior for zero length arrays is kept as a comment in case this can happen in the future.
EDIT: this change has been removed from this PR as a case where this could be possible was shown by @lavalamp. Instead some cleanup has been made in the same part: the else condition is not needed as delete is a no-op for non-existing keys.
2. A swallow copy of a slice is being done to instantly get the last element, thus making the copy useless. Issue #70751 tracks if a deep copy is needed, what would require a fix, but this PR keeps the current behavior.
3. Return the closed attribute directly instead of checking it and then returning a constant.
4. Reorder a counter decrease to stay near the element extraction

**Special notes for your reviewer**:
The first and last ones are mostly syntactic sugar, but the middle one should slightly improve the performance.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

roycaihw commented:
/cc

roycaihw commented:
/ok-to-test

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -320,19 +320,15 @@ func (f *DeltaFIFO) queueActionLocked(actionType DeltaType, obj interface{}) err
-		// in the queue are ignored if they are not in the
-		// map).
-		delete(f.items, id)
+	// newDeltas will always have at least one delta, if there is a case where
```
The intent of this code was that the deduping function could evolve over time. So, although it works this way now, changing the code makes it hard to optimize the deduping function.

Adirio replied:
Before suggesting this change I tried to think which implications it would have to remove this test. `dedupDeltas` may change over time but from a conceptual point of view, a dedup operation will only output an empty slice if the input is empty. This is the only asumption we are making for this change, and I think that the assumption is quite logical.

Adirio replied:
Makes sense, did not consider that option. Will remove that part tomorrow.

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -320,19 +320,15 @@ func (f *DeltaFIFO) queueActionLocked(actionType DeltaType, obj interface{}) err
-		// map).
-		delete(f.items, id)
+	// newDeltas will always have at least one delta, if there is a case where
+	// it may have no elements, delete f.items[id] (f.queue will ignore elements
```
This comment is confusingly worded, as it's talking about a case that is no longer handled here without making that clear.

Adirio replied:
Will rephrase if this change is accepted.

lavalamp commented:
sorry, I started reviewing this a while ago and got distracted.

I don't agree with the first change (I'm not arguing the existing code is optimal--it should at least be tested). The rest are fine.

Adirio commented:
> I don't agree with the first change (I'm not arguing the existing code is optimal--it should at least be tested). The rest are fine.

If you still think that we should test for empty slices, I will modify this PR to only include the two last points.

Adirio committed:
DeltaFIFO cleanup

Remove non-needed else condition
Remove non-needed swallow copy
Simplify return for IsClosed()
Keep amount decrement next to element extraction from the queue

Signed-off-by: Adrián Orive <adrian.orive.oneca@gmail.com>

Adirio commented:
@lavalamp Everything ready and checks passing (except the tide one as it requires LGTM, aproval and priority labels)

lavalamp commented on staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:
```diff
@@ -320,17 +320,15 @@ func (f *DeltaFIFO) queueActionLocked(actionType DeltaType, obj interface{}) err
-	_, exists := f.items[id]
 	if len(newDeltas) > 0 {
-		if !exists {
+		if _, exists := f.items[id]; !exists {
```
I probably wouldn't have made these changes at all but I guess I don't object :)

lavalamp replied:
(meaning, just these two--I'm a fan of the others)

lavalamp commented:
/lgtm
/approve
/milestone v1.13

Adirio commented:
I think it is just missing a priority now. Which one should I use?

lavalamp commented:
/priority important-soon

On Wed, Nov 14, 2018 at 3:32 PM Adrián <notifications@github.com> wrote:

> I think it is just missing a priority now. Which one should I use?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/70801#issuecomment-438858297>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAngltCvThMQD6S2vQRP3m5S2VXq0DPsks5uvKf5gaJpZM4YU2Vp>
> .
>
